### PR TITLE
fix flaky conformance test

### LINKED ISSUE #90506: [Flaky Test]: gce-master-scale-correctness (ci-kubernetes-e2e-gce-scale-correctness)
**Which jobs are failing**:
gce-master-scale-correctness (ci-kubernetes-e2e-gce-scale-correctness)

**Which test(s) are failing**:
[sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] works for CRD with validation schema [Conformance] 

**Since when has it been failing**:
04-26 05:02 PDT

**Testgrid link**:
https://testgrid.k8s.io/sig-release-master-informing#gce-master-scale-correctness

**Reason for failure**:
Apr 26 12:28:01.219: FAIL: failed to explain ksvc.spec: error running /workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://35.185.126.243 --kubeconfig=/workspace/.kube/config explain ksvc.spec:
Command stdout:

stderr:
the server doesn't have a resource type "ksvc"

error:
exit status 1

Full Stack Trace
k8s.io/kubernetes/test/e2e/apimachinery.glob..func4.1()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apimachinery/crd_publish_openapi.go:131 +0x1545
k8s.io/kubernetes/test/e2e.RunE2ETests(0xc002ca6100)
	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e.go:130 +0x324
k8s.io/kubernetes/test/e2e.TestE2E(0xc002ca6100)
	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e_test.go:141 +0x2b
testing.tRunner(0xc002ca6100, 0x4b8d060)
	/usr/local/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:960 +0x350
[AfterEach] [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:175
ï¿½[1mSTEPï¿½[0m: Collecting events from namespace "crd-publish-openapi-4377".
ï¿½[1mSTEPï¿½[0m: Found 0 events.
Apr 26 12:28:01.314: INFO: POD  NODE  PHASE  GRACE  CONDITIONS
Apr 26 12:28:01.314: INFO: 
Apr 26 12:28:02.899: INFO: skipping dumping cluster info - cluster too large
Apr 26 12:28:02.899: INFO: Waiting up to 3m0s for all (but 50) nodes to be ready
ï¿½[1mSTEPï¿½[0m: Destroying namespace "crd-publish-openapi-4377" for this suite.

**Anything else we need to know**:

/cc @kubernetes/ci-signal 
/cc @kubernetes/sig-api-machinery-bugs 
/milestone v1.19
/priority critical-urgent

### PR DESCRIPTION
**Pull request by knight42**
**What type of PR is this?**

/kind flake

**What this PR does / why we need it**:
Correct the flaky conformance test about `kubectl explain`.

**Which issue(s) this PR fixes**:

Fixes #90506 

**Special notes for your reviewer**:
Not sure where to place the non-conformace test, so I decided to simply remove it.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Review by hasheddan — APPROVED** [review]
/lgtm

**Review by soltysh — CHANGES_REQUESTED** [review]
I was under the impression we don't want to remove those bits, but rather move it to separate test which won't be part of conformance. Iow. you'll need to create a separate test exercising this functionality but don't add it to conformance.

**Comment by knight42** [comment]
@soltysh @liggitt PTAL

File: test/e2e/apimachinery/crd_publish_openapi.go [hunk 1]
```diff
@@ -126,22 +136,25 @@ var _ = SIGDescribe("CustomResourcePublishOpenAPI [Privileged:ClusterAdmin]", fu
...
-			framework.Failf("%v", err)
-		}
-
 		ginkgo.By("kubectl explain works to return error when explain is called on property that doesn't exist")
 		if _, err := framework.RunKubectl(f.Namespace.Name, "explain", crd.Crd.Spec.Names.Plural+".spec.bars2"); err == nil || !strings.Contains(err.Error(), `field "bars2" does not exist`) {
 			framework.Failf("unexpected no error when explaining property that doesn't exist: %v", err)
 		}
 
-		if err := cleanupCRD(f, crd); err != nil {
+		ginkgo.By("kubectl explain works for CR with the same resource name as built-in object.")
```
Comment (liggitt) [inline test/e2e/apimachinery/crd_publish_openapi.go:144]: this is still inside the conformance test

**Comment by liggitt** [comment]
this is still modifying a conformance test. please restore that test to the state it was in prior to https://github.com/kubernetes/kubernetes/pull/89505

**Comment by knight42** [comment]
@liggitt @soltysh 

I have found out that the root cause of the problem `the server doesn't have a resource type "ksvc"` is `CachedDiscoveryClient` trying to expand the short name `ksvc` using the stale local cache.

As far as I am concerned, we have two plans to solve this problem:
1. Force `CachedDiscoveryClient` to invalidate local cache just before https://github.com/kubernetes/client-go/blob/5159cff060fb98cd98592f6663f322c982066d7e/restmapper/shortcut.go#L87
2. Run `kubectl explain ksvc` twice in conformance test so that local cache can get refreshed.

Since Plan 1 may have influence that I cannot forsee, I chose Plan 2 here.

**Comment by liggitt** [comment]
that issue was tracked in https://github.com/kubernetes/kubernetes/issues/65517

I don't think it's a good idea to add workarounds for bugs into conformance tests like that...

**Comment by knight42** [comment]
@liggitt well yes... then what do you think of plan 1 I mentioned above?

**Comment by liggitt** [comment]
> @liggitt well yes... then what do you think of plan 1 I mentioned above?

Unconditional invalidation doesn't seem like a good idea. Let's get the flaky test isolated and out of conformance, then pursue fixing the issue in https://github.com/kubernetes/kubernetes/issues/65517

**Comment by knight42** [comment]
@liggitt PTAL now

**Comment by knight42** [comment]
spurious errors?

/retest

**Comment by knight42** [comment]
look like that we have bumped into other flaky tests?

**Pull request merged**